### PR TITLE
Add support to filter out CIDER's internal namespaces and vars

### PR DESCRIPTION
**Pull request by cskksc**
Before submitting a PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [ ] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [ ] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (`make test`)
- [ ] The new code is not generating bytecode or `M-x checkdoc` warnings
- [x] You've updated the changelog (if adding/changing user-visible functionality)
- [ ] You've updated the readme (if adding/changing user-visible functionality)
- [ ] You've updated the refcard (if you made changes to the commands listed there)

Thanks!

For #1564

### CODE REVIEW

File: cider.el [hunk 1]
```diff
@@ -204,6 +204,18 @@ Sub-match 1 must be the project path.")
 (defvar cider-host-history nil
   "Completion history for connection hosts.")
 
+(defcustom cider-filter-regexps '("cider.nrepl" "refactor-nrepl")
```
Comment (bbatsov) [inline cider.el]: As those are regexps, I guess they should start with `^` (beginning of string). I'd also filter nREPL's own namespaces.

File: cider.el [hunk 2]
```diff
@@ -204,6 +204,18 @@ Sub-match 1 must be the project path.")
...
+  "List of regexps used to filter out some internal details.
+When nil, nothing is filtered out. Otherwise, all namespaces matching any regexp
+from this list are dropped out of the `ns-list` op.
+Also, `apropos` won't include vars from such namespaces.
+This list is passed on to the nREPL middleware without any pre-processing.
+So the regexps have to be in Clojure format and not emacs-lisp."
```
Comment (bbatsov) [inline cider.el]: Emacs Lisp

File: doc/configuration.md [hunk 3]
```diff
@@ -112,6 +112,27 @@ To remove the prefix altogether just set it to an empty string(`""`).
 
 More details can be found [here](https://github.com/clojure-emacs/cider/issues/930).
 
+* You can hide all nREPL middleware details from `cider-browse-ns*` and `cider-apropos*`
+  commands by customizing the variable `cider-filter-regexps`. It should be a list of
+  regexps matching the pattern of namespaces you want to filter out.
+
+  Its default value is `'("cider.nrepl" "refactor-nrepl")`, the two most commonly
+  used middleware's.
```
Comment (bbatsov) [inline doc/configuration.md]: "middleware collections/packages" to be precise.

File: cider.el (hunk 2 above)
Comment (Malabarba) [inline cider.el]: > to be in Clojure format (with twice the number of backslashes)

File: cider.el (hunk 1 above)
Comment (bbatsov) [inline cider.el]: Not sure if we can't think of a better name for the variable. Right now it's not really clear what this is filtering without reading the docstring.

File: cider.el [hunk 4]
```diff
@@ -204,6 +204,18 @@ Sub-match 1 must be the project path.")
 (defvar cider-host-history nil
   "Completion history for connection hosts.")
 
+(defcustom cider-filter-regexps '("cider.nrepl" "refactor-nrepl")
+  "List of regexps used to filter out some internal details.
```
Comment (bbatsov) [inline cider.el]: some vars/symbols/namespaces

File: cider.el [hunk 5]
```diff
@@ -204,6 +204,18 @@ Sub-match 1 must be the project path.")
...
+When nil, nothing is filtered out. Otherwise, all namespaces matching any regexp
+from this list are dropped out of the `ns-list` op.
```
Comment (bbatsov) [inline cider.el]: Don't use backticks in Elisp docstrings. Use double quotes here.

File: doc/configuration.md [hunk 6]
```diff
@@ -112,6 +112,27 @@ To remove the prefix altogether just set it to an empty string(`""`).
...
+  To customize `cider-filter-regexps`, you could use the Emacs customize UI,
+  with <kbd>M-x</kbd> `customize-variable` <kbd>RET</kbd> `cider-filter-regexps`.
+
+  Or by including a similar snippet along with the other CIDER configuration.
+
+```el
+(setq cider-filter-regexps '(".*nrepl"))
+```
+
+
```
Comment (bbatsov) [inline doc/configuration.md]: Kill this extra blank line.

**Comment by bbatsov** [comment]
Why didn't you place this defcustom in `cider-client`? After all it's the only place its used.

File: cider.el (hunk 1 above)
Comment (Malabarba) [inline cider.el]: -filtered-namespaces-regexps?

Comment (bbatsov) [inline cider.el]: I'm not sure if this was applied just to the ns of something of to the entire fully qualified name.

**Comment by bbatsov** [comment]
You'll also have to rebase.

File: CHANGELOG.md [hunk 7]
```diff
@@ -29,6 +29,7 @@
 * Use `cider-apropos-select` instead of `cider-apropos` in `cider-apropos-documentation-select`.
 * [#1561](https://github.com/clojure-emacs/cider/issues/1561): Use an appropriate font-lock-face for variables, macros and functions in
 the ns-browser.
+* [#1564](https://github.com/clojure-emacs/cider/issues/1564): CIDER's internal namespaces and vars are filtered from the ns-browser and apropos functions.
```
Comment (bbatsov) [inline CHANGELOG.md]: Guess this should be in new features.

**Comment by cskksc** [comment]
I've made all the changes and rebased against the latest master. The `defcustom` is moved to `cider-client.el`; I'd kept it in `cider.el` as most of the custom vars are there. Also, the var is now named `cider-filtered-namespaces-regexps` as @Malabarba suggested; it is only being applied to the namespaces, returned by `(all-ns)`.

**Comment by bbatsov** [comment]
👍

**Pull request merged**